libcontainerd/local/local_windows.go

```diff
@@ -995,7 +995,7 @@ func (c *client) Pause(_ context.Context, containerID string) error {
 	}
 
 	if ctr.ociSpec.Windows.HyperV == nil {
-		return errors.New("cannot pause Windows Server Containers")
+		return errors.New("not implemented")
```

Comment:
I'm not clear why the text was changed (since the test also had its text changed... Looks like a no-op?) but would it be better to return [`errdefs.NotImplemented(errors.New("cannot pause Windows Server Containers"))`](https://github.com/moby/moby/blob/788f2883d285fc997d25cd27235c5b17cf3c0947/errdefs/helpers.go#L177-L183) and check using [`errdefs.IsNotImplemented`](https://github.com/moby/moby/blob/788f2883d285fc997d25cd27235c5b17cf3c0947/errdefs/is.go#L79-L83) in the test, so that the text is irrelevant?

Reply:
Ah, right, this is the local containerd API shim, I see why I couldn't find the other side of this change now.

Have you traced this through? I had a quick poke around, but I suspect this is coming from [hcsshim](https://github.com/microsoft/hcsshim/blob/v0.8.15/cmd/containerd-shim-runhcs-v1/service_internal.go#L252-L262) which is returning [`github.com/containerd/containerd/errdefs.ErrNotImplemented`](https://github.com/containerd/containerd/blob/v1.4.4/errdefs/errors.go#L49), and I'm not sure if the error *text* is part of the API, so it would probably be better for this code to return that same value, and the test can then test using [`github.com/containerd/containerd/errdefs.IsErrNotImplemented`](https://github.com/containerd/containerd/blob/v1.4.4/errdefs/errors.go#L79-L82) which makes us safe from any wrapping or spelling fixes that might happen.

Reply:
No. I just tried to get these tests working on both cases with minimal amount of changes. I would prefer go with my current proposal on this point because most probably Docker 21.x need support both runtimes (and deprecate old one). When HCS v1 removal starts these can be of course optimized.

Reply:
We would have to look what we want the error to be treated as; if we _detect_ the error from containerd as "not implemented" (using the containerd errdefs package), we can translate it to anything of course. If we _generate_ a `errdefs.NotImplemented()` using "our" errdefs package https://github.com/moby/moby/blob/788f2883d285fc997d25cd27235c5b17cf3c0947/errdefs/helpers.go#L177-L183, it will probably converted to a HTTP Notimplemented status in our API; https://github.com/moby/moby/blob/788f2883d285fc997d25cd27235c5b17cf3c0947/errdefs/http_helpers.go#L42-L43


So we need to check if that's the expected response code for this situation, or if it should be considered an `InvalidParameter`; https://github.com/moby/moby/blob/788f2883d285fc997d25cd27235c5b17cf3c0947/errdefs/helpers.go#L38

(which translates to a HTTP 400 return)

Reply:
Does this really matter that much? My plan was trying to help with this stuff as I had little bit spare time but I'm not sure if I'm really interested to start study most cleanest way to generate/check that error message.

Reply:
Right now, we send up either a raw error with some text (if "local" containerd) or something that is probably `github.com/containerd/containerd/errdefs.ErrNotImplemented` (from "remote" containerd's hcsshim runtime). Unless the code that calls this was handling it by matching the string (in which case this change is breaking that) then the observed behaviour shouldn't change.

The reason I suggested changing the test to match by type, and have this code send the same type as "remote" containerd does (i.e. changing the same two lines that are being changed now), is that the reason this fix is needed is that the last person to touch it also used string-matching for the test, and that broke... now.